runtime/parachains/src/paras.rs

```diff
@@ -524,9 +665,174 @@ pub mod pallet {
 			Self::deposit_event(Event::ActionQueued(para, next_session));
 			Ok(())
 		}
+
+		/// Includes a statement for a PVF pre-checking vote. Potentially, finalizes the vote and
+		/// enacts the results if that was the last vote before achieving the supermajority.
+		#[pallet::weight(0)]
+		pub fn include_pvf_check_statement(
+			origin: OriginFor<T>,
+			stmt: PvfCheckStatement,
+			signature: ValidatorSignature,
+		) -> DispatchResult {
+			ensure_none(origin)?;
+			let validators = shared::Pallet::<T>::active_validator_keys();
+			let current_session = shared::Pallet::<T>::session_index();
+			if stmt.session_index < current_session {
+				return Err(Error::<T>::PvfCheckStatementStale.into())
+			} else if stmt.session_index > current_session {
+				return Err(Error::<T>::PvfCheckStatementFuture.into())
+			}
+			let validator_index = stmt.validator_index.0 as usize;
+			let validator_public = validators
+				.get(validator_index)
+				.ok_or(Error::<T>::PvfCheckValidatorIndexOutOfBounds)?;
+
+			let signing_payload = stmt.signing_payload();
+			ensure!(
+				signature.verify(&signing_payload[..], &validator_public),
+				Error::<T>::PvfCheckInvalidSignature,
+			);
+
+			let mut active_vote = PvfActiveVoteMap::<T>::get(&stmt.subject)
+				.ok_or(Error::<T>::PvfCheckSubjectInvalid)?;
+
+			ensure!(
+				!active_vote
+					.has_vote(validator_index)
+					.ok_or(Error::<T>::PvfCheckValidatorIndexOutOfBounds)?,
+				Error::<T>::PvfCheckDoubleVote,
+			);
+
+			// Finally, cast the vote and persist.
+			if stmt.accept {
+				active_vote.votes_accept.set(validator_index, true);
+			} else {
+				active_vote.votes_reject.set(validator_index, true);
+			}
+
+			if let Some(outcome) = active_vote.quorum(validators.len()) {
+				// The supermajority quorum has been achieved.
+				//
+				// Remove the PVF vote from the active map and finalize the PVF checking according
+				// to the outcome.
+				PvfActiveVoteMap::<T>::remove(&stmt.subject);
+				PvfActiveVoteList::<T>::mutate(|l| {
+					if let Ok(i) = l.binary_search(&stmt.subject) {
+						l.remove(i);
+					}
+				});
```

Comment:
Reason to track it twice in a `Map` and `List` is entirely based on the perf of iterating over a `StorageMap`, correct?

Reply:
Yeah. I used the same approach when I was in a similar situation for messages. Arguably, there are more messages than PVF votes at the same time. However, I also think that it is simpler this way: it is easier to reason about the `Map` and `List` combination, IMO, than `iter_prefix`. Although I admit that it is a bit more verbose.

Reply:
Just making sure :)